kong/plugins/basic-auth/daos.lua

```diff
@@ -26,4 +27,20 @@ function BasicAuthCredentials:new(properties)
   BasicAuthCredentials.super.new(self, properties)
 end
 
+function BasicAuthCredentials:insert(params)
+  if params.password then
+    return BasicAuthCredentials.super.insert(self, params)
+  else
+    -- No password was provided, so we insert a random generated password
+    local newpwd = utils.random_string()
+    params.password = newpwd
+    local data, err = BasicAuthCredentials.super.insert(self, params)
+    -- inserting the data has encrypted the password field by now, 
+    -- so add a new field with the generated plain text password
+    -- which will be returned to the requester
+    if data then data.plain_password = newpwd end
+    return data, err
+  end
+end
```

Comment:
Okay, I'll explain in more details, as I don't know how much of this you already know, hang on. Sorry if I am repeating things you already know. Also, all of the following points are is still open to discussion, they are just the current implementation following the needs we faced with Kong (and the little time we had).

   One of Kong's requirements was to be database agnostic. In order to do that, we initially used to have a Model, represented with "Lua classes" for each entity, and a DAO, which was the actual implementation of a given database's operations. The application code would use the Model (with `insert/update...` methods), which in turn would call the DAO. This lead to a lot of duplicated core, and the addition of new entities was much more verbose, which we did not want because we wanted anybody to simply be able to define a new entity with a little effort as possible.

To do so, we decided to ditch out the Model layer. But since a common (and valid) design pattern wants one to implement validation logic at the Model layer, we came up with this stateless schema, which defines rules on an entity. Having the validation being performed on the Model allows it to be run when the entity is manipulated form anywhere (Admin API, CLI, Proxy logic...). The `schema_validations` and schemas work together to apply various rules on entities and properties of those entities. The schema validation is then called by the DAO before any DB operations.

In the long term, we would like several things:
- Kong to really be DB agnostic. That means an entity can be defined once, and any DAO implementation will follow the same validation rules. For now, the schema validation is called by the DAO itself, which is not optimal yet. When we implement the PostgreSQL DAO, it would have to also call the schema validation for example. This subject is still open to discussion.
- A developer could create an entity just with a schema definition. Currently this is not the case, a developer has to define the schema, **and** implement a DAO that simply inherits from the `base_dao` (since it implements basic CRUD operations and schema validations). implementing the DAO is most of the time just a formality, and the resulting DAO is simply inherited from `base_dao` without anything added. Like the `keyauth-credentials` DAO. But sometimes, one can add new methods (for example the APIs's DAO's `find_all()`, which are implemented differently depending on the database. PostgreSQL and Cassandra won't be executing the same code to retrieve all APIs, as you might have already guessed.

As you can see, a Kong entity is currently made of a schema, and a DAO. This design pattern is simpler than an ORM one (like we previously has with the Model classes) and, as we learned over time, more adapted for Lua and ngx_lua. Entities are simple tables, but rules apply on this table's values.

Hence, the real "Model" layer of Kong are the schemas. Now, I did not say the API to define and validate schema was complete. The `schema_validation` is where all the logic happens (it suffers from its own complexity, for example I prefer the validation I implemented for the config validation in `next`, which is more modular).

Back to the point of my initial comment: if you implement such logic in the DAO, this logic only applies for the Cassandra DAO, and not for any future DAO. It also prevents one from running, say, a dry-run of the entity validation. Which, we never know, might be useful for some validations in the Admin API or CLI. My idea was to follow the current design pattern of Kong: implement this logic in the key-credential's schema. Yes, that would probably mean improvements are to be made on the schema API (like you said,  `before_*` and `after_*` methods) (iirc, a `on_insert` method exists). The day we abstract DAO creation for entities, nobody will have to worry about re-implementing that logic in all existing DAO implementations (for each datastore I mean). It would also probably improve the schema validation API for other contributors and users (for example, #771 mentioned a way to "hide" fields, that has also been suggested in #385).

---

I hope I was clear enough. Maybe I will think of other things to add. The whole point is: schemas are the current Model layer of Kong. Validation should happen in the schema, and that the actual schema definition API is not yet finished and can certainly be cleaned-up/improved.